include/aws/s3/private/s3_buffer_pool.h

```diff
@@ -0,0 +1,96 @@
+#ifndef AWS_S3_BUFFER_ALLOCATOR_H
+#define AWS_S3_BUFFER_ALLOCATOR_H
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/s3/s3.h>
+
+/*
+ * S3 buffer pool.
+ * Buffer pool used for pooling part sized buffers for Put/Get.
+ * Provides additional functionally for limiting overall memory usage by setting
+ * upper bound beyond which buffer acquisition will fail.
+ */
+
+AWS_EXTERN_C_BEGIN
+
+struct aws_s3_buffer_pool;
+
+struct aws_s3_pooled_buffer {
+    size_t size;
+    uint8_t *ptr;
+};
```

Comment:
My reasoning for a new structure is that it makes it obvious that we should clean it up on its own - it felt really weird passing regular buf or cursor back to buffer pool.

i put struct into header since its private and its easier to deal with it by value, but users should not care about whats in it. and later on we can add fancier things to it for making pooling easier, like which block it came from

Reply:
trivial: aws_byte_cursor seems fine to me